### ISSUE TITLE
Inject Noise in Training Samples in an Online Fashion

### ISSUE THREAD

**Issue by pGit1** [aigamedev/scikit-neuralnetwork#196]
Hi, 

I was wondering if there was a way to inject masking noise (or any other distortion) in an online fashion while training the net. Currently I add noise to the training set prior to training but if samples could be mutated at the end of each training epoch for instance or the end of a batch pass that would be awesome. This package is awesome! Im not saying it cant do it, I just need a little help figuring it out.

**Comment by alexjc** [comment]
You can do two things:
- Use the `callback` for `on_batch_start` to mutate your training samples.
- Use the new `Native` layers to insert a gaussian noise layer directly from lasagne.

Both are in the documentation, but if you need help or want to include this specific example in the docs, let me know!

**Comment by pGit1** [comment]
Alex I would love a specific example in the docs, it would be so helpful! I
think a callback example specifically would be awesome. I haven't tested
the Native functionality yet so I think that would be helpful as well.

I think the main part I need help with is with the function to create
masking noise and Gaussian noise for the training samples.

Thanks for your help!

**Comment by alexjc** [comment]
Native layers are documented already, see tests for a `GaussianNoiseLayer`.
http://scikit-neuralnetwork.readthedocs.org/en/latest/guide_model.html#native-custom-layers

Documentation also has an example for callbacks of course.

**Comment by pGit1** [comment]
Alex, so both methods cause the noise to be injected in an online fashion?
I THINK I got the callback method to work. I am testing it now and it is
taking a LONG time to produce results on a simple test network. Perhaps I
am doing something wrong. Some code below:

def corrupt_data(X, **kwargs):
    keep_prob = np.random.rand()
    mask_noise = np.random.binomial(n=1,p=keep_prob,size=X.shape)
    X = X*mask_noise

callback={'on_batch_start': corrupt_data}

**Comment by alexjc** [comment]
Hrm, for `corrupt_data` you need to process `Xb` so it works on each batch. Also, you need to update to the same vector, so `Xb[:] = Xb*mask_noise`. However, this would likely destroy your dataset so you'd have to reload it each time, or reset it `on_batch_finish`.

**Comment by pGit1** [comment]
What would my function look like? I am not sure how or where to make the
modification. Also why do you think my function took so long to process
when doing the callback during fitting?

**Comment by alexjc** [comment]
It took so long because you modified all `X` and not the batch `Xb`. I'm not sure there's an easy function that would work, you'd have to store a copy of your data. Use the lasagne layer for gaussian noise...

**Comment by pGit1** [comment]
Ok thanks Alex! Ive been playing with Native so I think this should work
fine.

**Comment by pGit1** [comment]
Alex can you show me an example of working gaussiannoiselayer code? I can't
seem to get it ti work...

**Comment by alexjc** [comment]
I've never done it, but see `test_native.py` and find `GaussianNoiseLayer`.

**Comment by pGit1** [comment]
Ok so based on the testing code I can get it to run but keep getting this
error now:

'

ERROR:sknn:A runtime exception was caught during training. This likely
occurred due to
a divergence of the SGD algorithm, and NaN floats were found by the backend.

Try setting the `learning_rate` 10x lower to resolve this, for example:
    learning_rate=0.000000


'

The callback seems to be a better fit? Not sure what in the implementation
would require the learning rate to be decreased so much just by adding
noise. This is not the case when using a callback function.

**Comment by pGit1** [comment]
Also will this work to add noise and reset the data appropriately. Im trying to work around the native issue:

`def inject_noise(Xb,**others):
    gauss_noise = np.random.normal(0,1,size=Xb.shape)
    Xb[:] = Xb+gauss_noise
    
def reset_data(X,**others):
    X[:] = X` 

and 

`callback={'on_batch_start': inject_noise,'on_epoch_finish': reset_data}`

The network works fine when I run this code but I wonder if this is actually corrupting each batch before learning and then resetting ALL the data (batches) back to normal before the next epoch. 

Thanks for your help Alex.
